implementations.tex

```diff
@@ -52,10 +52,8 @@ \section{Execution Based} \label{execution_based}
 When {\tt ebreak} is executed (indicating the end of the
 Program Buffer code) the hart returns to its park loop. If an exception is
 encountered, the hart jumps to a debug trap address within
-the Debug Module. The code at that address causes the hart to
-write to an address in the Debug Module which indicates exception.
-This address is considered I/O for {\tt fence} instructions (see \#\ref{fence}
-on page \pageref{fence}).
+the Debug Module. The code there causes the hart to
+write to the Debug Module indicating an exception.
 Then the hart jumps back to the park loop.
 The DM infers from the write that there was an exception, and sets \FdmAbstractcsCmderr appropriately.
```

Comment:
```suggestion
the Debug Module. The code there causes the hart to
write to the Debug Module indicating an exception.
Then the hart jumps back to the park loop.
The DM infers from the write that there was an exception, and sets \FdmAbstractcsCmderr appropriately.
Before entering the park loop, the hart will typically execute a {\tt fence} instruction, then write to the Debug Module indicating that the abstract command has finished.  The {\tt fence} ensures that any effects from the abstract command take effect before the DM returns \FdmAbstractcsBusy to 0.
```

Reply:
That's a good starting point. Thanks. I was going to extend it by saying that the fence should happen both in the park loop and the exception handler, but rocket-chip does not have a fence at the start of the exception handler: https://github.com/chipsalliance/rocket-chip/blob/master/scripts/debug_rom/debug_rom.S#L64

The comment on the `fence` in _entry also doesn't match what we're saying here.

I think the language we're writing here is correct, and rocket-chip either isn't, or has some hack that's not visible in debug_rom.S to make things work out.

Reply:
A rocket-chip exception writes to EXCEPTION, then does an ebreak which goes back to _entry, which then executes the fence.  The rocket-chip comment is just saying that if the abstract command writes something to DATA (such as a register read value), the fence ensures that the value in DATA is visible before the write to HALTED clears the abstractcs.busy bit.  Without the fence, conceivably the write to DATA could be delayed relative to HALTED.

Reply:
Got it. Thanks for explaining.